File: java/com/joshmanisdabomb/lcc/entity/render/StateBasedTNTRenderer.java
package com.joshmanisdabomb.lcc.entity.render;

import com.joshmanisdabomb.lcc.misc.AdaptedFromSource;
import com.mojang.blaze3d.matrix.MatrixStack;
import net.minecraft.block.BlockState;
import net.minecraft.client.renderer.IRenderTypeBuffer;
import net.minecraft.client.renderer.Vector3f;
import net.minecraft.client.renderer.entity.EntityRendererManager;
import net.minecraft.client.renderer.entity.TNTMinecartRenderer;
import net.minecraft.client.renderer.entity.TNTRenderer;
import net.minecraft.entity.item.TNTEntity;
import net.minecraft.util.math.MathHelper;

public class StateBasedTNTRenderer extends TNTRenderer {

    private final BlockState state;

    public StateBasedTNTRenderer(BlockState state, EntityRendererManager manager) {
        super(manager);
        this.state = state;
    }

    @Override
    @AdaptedFromSource
    public void render(TNTEntity entityIn, float entityYaw, float partialTicks, MatrixStack matrixStackIn, IRenderTypeBuffer bufferIn, int packedLightIn) {
        matrixStackIn.push();
        matrixStackIn.translate(0.0D, 0.5D, 0.0D);
        if ((float)entityIn.getFuse() - partialTicks + 1.0F < 10.0F) {
            float f = 1.0F - ((float)entityIn.getFuse() - partialTicks + 1.0F) / 10.0F;
            f = MathHelper.clamp(f, 0.0F, 1.0F);
            f = f * f;
            f = f * f;
            float f1 = 1.0F + f * 0.3F;
            matrixStackIn.scale(f1, f1, f1);
        }

        matrixStackIn.rotate(Vector3f.YP.rotationDegrees(-90.0F));
        matrixStackIn.translate(-0.5D, -0.5D, 0.5D);
        matrixStackIn.rotate(Vector3f.YP.rotationDegrees(90.0F));
        TNTMinecartRenderer.renderTntFlash(state, matrixStackIn, bufferIn, packedLightIn, entityIn.getFuse() / 5 % 2 == 0);
        matrixStackIn.pop();

        //super.super.render
        net.minecraftforge.client.event.RenderNameplateEvent renderNameplateEvent = new net.minecraftforge.client.event.RenderNameplateEvent(entityIn,entityIn.getDisplayName().getFormattedText(), matrixStackIn, bufferIn);
        net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(renderNameplateEvent);
        if (renderNameplateEvent.getResult() != net.minecraftforge.eventbus.api.Event.Result.DENY && (renderNameplateEvent.getResult() == net.minecraftforge.eventbus.api.Event.Result.ALLOW || this.canRenderName(entityIn))) {
            this.renderName(entityIn, renderNameplateEvent.getContent(), matrixStackIn, bufferIn, packedLightIn);
        }
    }

}

File: kotlin/com/joshmanisdabomb/lcc/entity/render/StateBasedTNTRenderer.kt
package com.joshmanisdabomb.lcc.entity.render

import net.minecraft.block.BlockState
import net.minecraft.client.render.VertexConsumerProvider
import net.minecraft.client.render.entity.EntityRendererFactory
import net.minecraft.client.render.entity.TntEntityRenderer
import net.minecraft.client.render.entity.TntMinecartEntityRenderer
import net.minecraft.client.util.math.MatrixStack
import net.minecraft.entity.TntEntity
import net.minecraft.util.math.MathHelper
import net.minecraft.util.math.Vec3f

class StateBasedTNTRenderer(val state: BlockState, dispatcher: EntityRendererFactory.Context) : TntEntityRenderer(dispatcher) {

    override fun render(entity: TntEntity, yaw: Float, tickDelta: Float, matrixStack: MatrixStack, vertexConsumerProvider: VertexConsumerProvider, light: Int) {
        matrixStack.push()
        matrixStack.translate(0.0, 0.5, 0.0)
        if (entity.fuse.toFloat() - tickDelta + 1.0f < 10.0f) {
            var h = 1.0f - (entity.fuse.toFloat() - tickDelta + 1.0f) / 10.0f
            h = MathHelper.clamp(h, 0.0f, 1.0f)
            h *= h
            h *= h
            val k = 1.0f + h * 0.3f
            matrixStack.scale(k, k, k)
        }
        matrixStack.multiply(Vec3f.POSITIVE_Y.getDegreesQuaternion(-90.0f))
        matrixStack.translate(-0.5, -0.5, 0.5)
        matrixStack.multiply(Vec3f.POSITIVE_Y.getDegreesQuaternion(90.0f))
        TntMinecartEntityRenderer.renderFlashingBlock(state, matrixStack, vertexConsumerProvider, light, entity.fuse / 5 % 2 == 0)
        matrixStack.pop()
        if (hasLabel(entity)) {
            renderLabelIfPresent(entity, entity.displayName, matrixStack, vertexConsumerProvider, light)
        }
    }

}
